src/hotspot/share/asm/register.hpp

```diff
@@ -93,67 +93,75 @@ template <class RegImpl> class ReverseRegSetIterator;
 // A set of registers
 template <class RegImpl>
 class AbstractRegSet {
-  uint32_t _bitset;
+  size_t _bitset;
 
-  AbstractRegSet(uint32_t bitset) : _bitset(bitset) { }
+  constexpr AbstractRegSet(size_t bitset) : _bitset(bitset) { }
+
+  static constexpr int max_size() {
+    return (int)(sizeof _bitset * CHAR_BIT);
+  }
 
 public:
 
-  AbstractRegSet() : _bitset(0) { }
+  constexpr AbstractRegSet() : _bitset(0) { }
 
-  AbstractRegSet(RegImpl r1) : _bitset(1 << r1->encoding()) { }
+  constexpr AbstractRegSet(RegImpl r1)
+    : _bitset(r1->is_valid() ? size_t(1) << r1->encoding() : 0) { }
 
-  AbstractRegSet operator+(const AbstractRegSet aSet) const {
+  constexpr AbstractRegSet operator+(const AbstractRegSet aSet) const {
     AbstractRegSet result(_bitset | aSet._bitset);
     return result;
   }
 
-  AbstractRegSet operator-(const AbstractRegSet aSet) const {
+  constexpr AbstractRegSet operator-(const AbstractRegSet aSet) const {
     AbstractRegSet result(_bitset & ~aSet._bitset);
     return result;
   }
 
-  AbstractRegSet &operator+=(const AbstractRegSet aSet) {
+  constexpr AbstractRegSet &operator+=(const AbstractRegSet aSet) {
     *this = *this + aSet;
     return *this;
   }
 
-  AbstractRegSet &operator-=(const AbstractRegSet aSet) {
+  constexpr AbstractRegSet &operator-=(const AbstractRegSet aSet) {
     *this = *this - aSet;
     return *this;
   }
 
-  static AbstractRegSet of(RegImpl r1) {
+  constexpr static AbstractRegSet of(RegImpl r1) {
     return AbstractRegSet(r1);
   }
 
-  static AbstractRegSet of(RegImpl r1, RegImpl r2) {
+  constexpr static AbstractRegSet of(RegImpl r1, RegImpl r2) {
     return of(r1) + r2;
   }
 
-  static AbstractRegSet of(RegImpl r1, RegImpl r2, RegImpl r3) {
+  constexpr static AbstractRegSet of(RegImpl r1, RegImpl r2, RegImpl r3) {
     return of(r1, r2) + r3;
   }
 
-  static AbstractRegSet of(RegImpl r1, RegImpl r2, RegImpl r3, RegImpl r4) {
+  constexpr static AbstractRegSet of(RegImpl r1, RegImpl r2, RegImpl r3, RegImpl r4) {
     return of(r1, r2, r3) + r4;
   }
 
-  static AbstractRegSet range(RegImpl start, RegImpl end) {
+  constexpr static AbstractRegSet range(RegImpl start, RegImpl end) {
     int start_enc = start->encoding();
     int   end_enc = end->encoding();
     assert(start_enc <= end_enc, "must be");
-    uint32_t bits = ~0;
+    size_t bits = ~(size_t)0;
     bits <<= start_enc;
-    bits <<= 31 - end_enc;
-    bits >>= 31 - end_enc;
+    bits <<= max_size() - 1 - end_enc;
+    bits >>= max_size() - 1 - end_enc;
 
     return AbstractRegSet(bits);
   }
 
-  uint size() const { return population_count(_bitset); }
+  constexpr bool contains(RegImpl reg) {
+    return (AbstractRegSet(reg).bits() & bits()) != 0;
+  }
 
-  uint32_t bits() const { return _bitset; }
+  constexpr uint size() const { return population_count(_bitset); }
```

Comment:
population_count is currently not constexpr.  I'm surprised this doesn't lead to warnings for guaranteed
non-constexpr body of a constexpr function.  I'm pretty sure I've seen such warnings from some compiler.
Note that I think there's no reason for it to not be constexpr, other than preceding our use of C++11/14 and
not yet updated accordingly.  https://bugs.openjdk.org/browse/JDK-8323952